### PR TITLE
Add python-level structured config

### PR DESCRIPTION
**Pull request by deepkyu**
## Description

Please include a summary in English, of the changes in this pull request. If it closes an issue, please mention it here.

Closes: #203

You should link at least one existing issue for PR. Before your create a PR, please check to see if there is an issue for this change.  
PRs from forked repository not accepted.

## Change(s)

- Add python-level structured config which is 100% compatible with the current yaml configuration
  - User can train the model manipulating with pythonic configuration
  - Backward-support: `netspresso-train **yaml_files` on CLI interface also works
  - Pythonic configuration can be also exported to yaml files when saving experiment log

https://github.com/Nota-NetsPresso/netspresso-trainer/blob/717b6ff24693ffd20d4e542fee7cd44410a1fd6d/train.py#L4-L38

## Changelog

If you PR to `dev` branch, please add a brief summary of the change to the **Upcoming Release** section of the [`CHANGELOG.md`](https://github.com/Nota-NetsPresso/netspresso-trainer/blob/master/CHANGELOG.md) file and include a link to the PR (formatted in markdown) and a link to your github profile.

For example,

```
- Added a new feature by `@myusername` in [PR 2023](https://github.com/Nota-NetsPresso/netspresso-trainer/pull/2023)
```

## Code Formatting

If you PR to either `master` or `dev` branch, you should follow the code linting process. Please check your code with `lint_check.sh` in `./scripts` directory.
For more information, please read the contribution guide in `CONTRIBUTING.md`.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -1,5 +1,40 @@
...
+    # ###
+    
+    
+    """
+    Declare trainer config
+    """
+    cfg = TrainerConfig(
+        task='classification',
+        auto=True,
+        data=ExampleBeansDataset,
```
Comment (illian01) [inline train.py:28]: I think `data` parameter should use `example_dataset` variable defined at line 9.

**Review by illian01 — CHANGES_REQUESTED** [review]
Thank you for your great works. :+1:
I found minor mismatch. Please check the comment.

File: train.py (hunk 1 above)
Comment (deepkyu) [reply train.py]: Thx. I fixed it.

**Review by illian01 — APPROVED** [review]
LGTM!

File: src/netspresso_trainer/cfg/training.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+from dataclasses import dataclass
+
+from omegaconf import MISSING, MissingMandatoryValue
+
+
+@dataclass
+class ScheduleConfig:
+    seed: int = 1
+    opt: str = "adamw"
```
Comment (Only-bottle) [inline src/netspresso_trainer/cfg/training.py:9]: P4) I think it would be good to handle optimizer & scheduler as Enum, what do you think? ([Optimizer support list in np-trainer docs](https://nota-netspresso.github.io/netspresso-trainer/components/training/#supporting-optimizers)) 

```
ex)
Optimizer.Adamw
Optimizer.Adam
...

Scheduler.Cosine
...
```

**Review by Only-bottle — APPROVED** [review]

File: src/netspresso_trainer/cfg/training.py (hunk 2 above)
Comment (deepkyu) [reply src/netspresso_trainer/cfg/training.py:9]: @Only-bottle 

We may internally construct `Enum`, but I personally believe that it doesn't need to be exposed to the user.
Most training repositories just keep it as string nickname in their training script.
It might also be the problem if the user uses trainer with the yaml configuration, which does not support pythonic `Enum` itself.

Comment (deepkyu) [reply src/netspresso_trainer/cfg/training.py:9]: @illian01 How do you think?

Comment (deepkyu) [reply src/netspresso_trainer/cfg/training.py:9]: Insider discussion: no change, keep it as is.
